src/renderer/atlas/AtlasEngine.r.cpp

```diff
@@ -540,64 +537,436 @@ void AtlasEngine::_drawCursor()
+{
+    __assume(d2dRenderTarget != nullptr);
+
+    if (scalingRequired)
+    {
+        const D2D1_MATRIX_3X2_F transform{
+            scale.x,
+            0,
+            0,
+            scale.y,
+            (origin.x + halfSize.x) * (1.0f - scale.x),
+            (origin.y + halfSize.y) * (1.0f - scale.y),
+        };
+        d2dRenderTarget->SetTransform(&transform);
+    }
+}
+
+void AtlasEngine::CachedGlyphLayout::undoScaling(ID2D1RenderTarget* d2dRenderTarget) const noexcept
+{
+    __assume(d2dRenderTarget != nullptr);
+
+    if (scalingRequired)
+    {
+        static constexpr D2D1_MATRIX_3X2_F identity{ 1, 0, 0, 1, 0, 0 };
+        d2dRenderTarget->SetTransform(&identity);
+    }
+}
+
+void AtlasEngine::_d2dPresent()
+{
+    auto dirtyRectInPx = _r.dirtyRect;
+    dirtyRectInPx.left *= _r.fontMetrics.cellSize.x;
+    dirtyRectInPx.top *= _r.fontMetrics.cellSize.y;
+    dirtyRectInPx.right *= _r.fontMetrics.cellSize.x;
+    dirtyRectInPx.bottom *= _r.fontMetrics.cellSize.y;
+
+    if (const auto intersection = _r.previousDirtyRectInPx & dirtyRectInPx)
+    {
+        wil::com_ptr<ID3D11Resource> backBuffer;
+        wil::com_ptr<ID3D11Resource> frontBuffer;
+        THROW_IF_FAILED(_r.swapChain->GetBuffer(0, __uuidof(backBuffer), backBuffer.put_void()));
+        THROW_IF_FAILED(_r.swapChain->GetBuffer(1, __uuidof(frontBuffer), frontBuffer.put_void()));
+
+        D3D11_BOX intersectBox;
+        intersectBox.left = intersection.left;
+        intersectBox.top = intersection.top;
+        intersectBox.front = 0;
+        intersectBox.right = intersection.right;
+        intersectBox.bottom = intersection.bottom;
+        intersectBox.back = 1;
+        _r.deviceContext->CopySubresourceRegion1(backBuffer.get(), 0, intersection.left, intersection.top, 0, frontBuffer.get(), 0, &intersectBox, 0);
+    }
+
+    _d2dCreateRenderTarget();
+    _d2dDrawDirtyArea();
+
+    // See documentation for IDXGISwapChain2::GetFrameLatencyWaitableObject method:
+    // > For every frame it renders, the app should wait on this handle before starting any rendering operations.
+    // > Note that this requirement includes the first frame the app renders with the swap chain.
+    assert(debugGeneralPerformance || _r.frameLatencyWaitableObjectUsed);
+
+    if (dirtyRectInPx)
+    {
+        {
+            RECT scrollRect{};
+            POINT scrollOffset{};
+            DXGI_PRESENT_PARAMETERS params{
+                .DirtyRectsCount = 1,
+                .pDirtyRects = dirtyRectInPx.as_win32_rect(),
+            };
+
+            if (_r.scrollOffset)
+            {
+                scrollRect = {
+                    0,
+                    std::max<til::CoordType>(0, _r.scrollOffset),
+                    _r.cellCount.x,
+                    _r.cellCount.y + std::min<til::CoordType>(0, _r.scrollOffset),
+                };
+                scrollOffset = {
+                    0,
+                    _r.scrollOffset,

```

Comment:
we change `previousDirtyRectInPx`, but we don't clear `scrollOffset`. Is this OK?

Reply:
I believe so. `previousDirtyRectInPx` is the only thing that should carry over to the next frame.